atc/db/resource_check_rate_limiter.go

```diff
@@ -0,0 +1,105 @@
+package db
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"time"
+
+	"code.cloudfoundry.org/clock"
+	"golang.org/x/time/rate"
+)
+
+type ResourceCheckRateLimiter struct {
+	clock clock.Clock
+
+	checkInterval time.Duration
+	checkLimiter  *rate.Limiter
+
+	refreshConn    Conn
+	refreshLimiter *rate.Limiter
+
+	mut *sync.Mutex
+}
+
+func NewResourceCheckRateLimiter(
+	refreshConn Conn,
+	checkInterval time.Duration,
+	refreshInterval time.Duration,
+	clock clock.Clock,
+) *ResourceCheckRateLimiter {
+	return &ResourceCheckRateLimiter{
+		clock: clock,
+
+		checkInterval: checkInterval,
+		checkLimiter:  nil,
+
+		refreshConn:    refreshConn,
+		refreshLimiter: rate.NewLimiter(rate.Every(refreshInterval), 1),
+
+		mut: new(sync.Mutex),
+	}
+}
+
+func (limiter *ResourceCheckRateLimiter) Wait(ctx context.Context) error {
+	limiter.mut.Lock()
+	defer limiter.mut.Unlock()
```

Comment:
Is holding the lock while we wait going to cause issues while we wait for the `delay` to elapse? I suppose if if any check is rate limited, no other check will be able to run until at least `delay` passes, so perhaps it's fine - but curious about whether your reasoning was the same